src/PureSMT/Process.hs

```diff
@@ -30,14 +29,18 @@ initZ3Instance ::
   Bool ->
   IO Solver
 initZ3Instance dbg = do
+  let ctxFinalizer =
+        [C.funPtr| void free_context(Z3_context ctx) {
+                                      Z3_del_context(ctx);
+                                      } |]
   solverCtx <-
-    [CU.block| Z3_context {
+    newForeignPtr ctxFinalizer
```

Comment:
The finalizer can run in a different thread that the one that created the context. But perhaps that's fine for Z3. When you test it, you would have to make sure to build the test with the multi-threaded runtime.